keyvi/include/keyvi/dictionary/dictionary_compiler.h

```diff
@@ -250,15 +162,183 @@ class DictionaryCompiler final {
   }
 
  private:
-  SorterT sorter_;
   keyvi::util::parameters_t params_;
+  key_values_t key_values_;
   ValueStoreT* value_store_;
   typename GeneratorAdapter::AdapterPtr generator_;
   std::string manifest_;
-  size_t count_ = 0;
+  size_t memory_limit_;
+  size_t memory_estimate_ = 0;
+  size_t chunk_ = 0;
   size_t size_of_keys_ = 0;
-  bool sort_finalized_ = false;
-  bool stable_insert_ = false;
+  size_t parallel_sort_threshold_ = 10000;
+  boost::filesystem::path temporary_directory_;
+
+  inline void Sort() {
+    if (key_values_.size() > parallel_sort_threshold_ && parallel_sort_threshold_ != 0) {
+      boost::sort::block_indirect_sort(key_values_.begin(), key_values_.end());
+    } else {
+      std::sort(key_values_.begin(), key_values_.end());
+    }
+  }
+
+  inline void TriggerSortAndChunkGenerationIfNeeded() {
+    if (memory_estimate_ < memory_limit_) {
+      return;
+    }
+    CreateChunk();
+  }
+
+  inline void CreateChunk() {
+    TRACE("create chunk %ul", key_values_.size());
+    if (chunk_ == 0) {
+      boost::filesystem::create_directory(temporary_directory_);
+    }
+
+    Sort();
+
+    // disable minimization for faster compile
+    keyvi::util::parameters_t params(params_);
+    params[MINIMIZATION_KEY] = "off";
+    fsa::Generator<keyvi::dictionary::fsa::internal::SparseArrayPersistence<uint16_t>, fsa::internal::NullValueStore,
+                   uint32_t, int32_t>
+        generator(params);
+
+    for (const key_value_t& key_value : key_values_) {
+      TRACE("adding to generator: %s", key_value.key.c_str());
+      generator.Add(std::move(key_value.key), key_value.value);
```

Comment:
`std::move()` here does make a copy, as  `key_value` is `const`, better to directly pass `key_value.key`, so it can go by ref